httpbp/middlewares.go

```diff
@@ -299,3 +306,105 @@ func recoverPanic(name string, next HandlerFunc) HandlerFunc {
 		return next(ctx, w, r)
 	}
 }
+
+// statusCodeRecorder is used by RecordStatusCode to record the code passed to
+// a call to WriteHeader.
+type statusCodeRecorder struct {
+	http.ResponseWriter
+
+	code int
+}
+
+func (r *statusCodeRecorder) WriteHeader(code int) {
+	r.ResponseWriter.WriteHeader(code)
+	if r.code == 0 {
+		r.code = code
+	}
+}
+
+func (r *statusCodeRecorder) getCode(err error) int {
+	if r.code != 0 {
+		// WriteHeader was called explicitly, use that
+		return r.code
+	}
+	if err != nil {
+		// something went wrong, check if err is an HTTPErr where we can extract
+		// the code, otherwise assume InternalServerError
+		var httpErr HTTPError
+		if errors.As(err, &httpErr) {
+			return httpErr.Response().Code
+		}
+		return http.StatusInternalServerError
+
+	}
+	// if there's no error returned and no call to WriteHeader, Go will
+	// return OK.
+	// https://pkg.go.dev/net/http#ResponseWriter.WriteHeader
+	return http.StatusOK
+}
+
+var families = [10]string{
+	"nan",
+	"1xx",
+	"2xx",
+	"3xx",
+	"4xx",
+	"5xx",
+	"6xx",
```

Comment:
🔕  there's nothing in the rfc currently for values >= 600. now with the table approach I'd suggest we just put everything >= 600 into nan as well (we can add new codes when they are added to the rfc)

Reply:
Ok, I don't feel too strong either way, I just went up to `9` because `WriteHeader` lets you write anything between 100 and 999.